File: packages/han/lib/data/hosted/schema/native-tasks.ts
/**
 * Native Tasks Schema
 *
 * Tracks Claude Code's built-in task system (TaskCreate/TaskUpdate).
 */

import {
	integer,
	pgTable,
	text,
	timestamp,
	unique,
	uuid,
	varchar,
} from "drizzle-orm/pg-core";
import { organizations } from "./organizations.ts";
import { sessions } from "./sessions.ts";

/**
 * Native Tasks table
 *
 * Stores tasks created via Claude Code's TaskCreate/TaskUpdate tools.
 */
export const nativeTasks = pgTable(
	"native_tasks",
	{
		id: uuid("id").primaryKey().defaultRandom(),
		organizationId: uuid("organization_id")
			.notNull()
			.references(() => organizations.id, { onDelete: "cascade" }),
		sessionId: uuid("session_id")
			.notNull()
			.references(() => sessions.id, { onDelete: "cascade" }),
		// The task ID from Claude (e.g., "1", "2") - scoped to session
		localTaskId: varchar("local_task_id", { length: 50 }).notNull(),
		messageId: uuid("message_id"), // References the message that created/updated this task
		subject: text("subject").notNull(),
		description: text("description"),
		status: varchar("status", { length: 50 }).notNull().default("pending"), // 'pending', 'in_progress', 'completed'
		activeForm: text("active_form"), // Present continuous form
		owner: varchar("owner", { length: 255 }),
		blocks: text("blocks"), // JSON array of task IDs
		blockedBy: text("blocked_by"), // JSON array of task IDs
		lineNumber: integer("line_number").notNull(),
		createdAt: timestamp("created_at", { withTimezone: true }).notNull(),
		updatedAt: timestamp("updated_at", { withTimezone: true }).notNull(),
		completedAt: timestamp("completed_at", { withTimezone: true }),
	},
	(table) => [unique().on(table.sessionId, table.localTaskId)],
);

/**
 * TypeScript type for a native task row
 */
export type NativeTask = typeof nativeTasks.$inferSelect;
export type NewNativeTask = typeof nativeTasks.$inferInsert;

File: packages/han-rs/crates/han-db/src/entities/native_tasks.rs
//! Entity: native_tasks (from Claude Code's TaskCreate/TaskUpdate tools)

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel, Serialize, Deserialize)]
#[sea_orm(table_name = "native_tasks")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    pub id: String,
    pub session_id: String,
    pub message_id: String,
    pub subject: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    pub status: String,
    pub active_form: Option<String>,
    pub owner: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub blocks: Option<String>,
    #[sea_orm(column_type = "Text", nullable)]
    pub blocked_by: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub completed_at: Option<String>,
    pub line_number: i32,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {
    #[sea_orm(
        belongs_to = "super::sessions::Entity",
        from = "Column::SessionId",
        to = "super::sessions::Column::Id"
    )]
    Session,
}

impl Related<super::sessions::Entity> for Entity {
    fn to() -> RelationDef {
        Relation::Session.def()
    }
}

impl ActiveModelBehavior for ActiveModel {}
